src/contract_wrappers/exchange_wrapper.ts

```diff
@@ -147,28 +147,41 @@ export class ExchangeWrapper extends ContractWrapper {
     public async fillOrderAsync(signedOrder: SignedOrder, fillTakerTokenAmount: BigNumber.BigNumber,
                                 shouldThrowOnInsufficientBalanceOrAllowance: boolean,
                                 takerAddress: string): Promise<BigNumber.BigNumber> {
+        // validate input types
         assert.doesConformToSchema('signedOrder', signedOrder, schemas.signedOrderSchema);
         assert.isBigNumber('fillTakerTokenAmount', fillTakerTokenAmount);
         assert.isBoolean('shouldThrowOnInsufficientBalanceOrAllowance', shouldThrowOnInsufficientBalanceOrAllowance);
-        await assert.isSenderAddressAsync('takerAddress', takerAddress, this._web3Wrapper);
-
-        const exchangeInstance = await this._getExchangeContractAsync();
-        await this.validateFillOrderThrowIfInvalidAsync(signedOrder, fillTakerTokenAmount, takerAddress);
 
+        // extract orderAdresses and orderValues from the provided order
         const [orderAddresses, orderValues] = ExchangeWrapper._getOrderAddressesAndValues(signedOrder);
 
-        const gas = await exchangeInstance.fillOrder.estimateGas(
-            orderAddresses,
-            orderValues,
-            fillTakerTokenAmount,
-            shouldThrowOnInsufficientBalanceOrAllowance,
-            signedOrder.ecSignature.v,
-            signedOrder.ecSignature.r,
-            signedOrder.ecSignature.s,
-            {
-                from: takerAddress,
-            },
-        );
+        // build a function that returns a promise that makes an RPC to retrieve the Exchange contract (may be memoized), 
+        // and then makes and RPC to estimate gas cost with the Exchange contract instance
+        const getExchangeInstanceAndGasAsync = async (): Promise<[ExchangeContract, number]> => {
+            const exchangeInstance = await this._getExchangeContractAsync();
```

Comment:
This method is already memoized and returns the exchange contract instance if it's already been instantiated. This is also used by the `validateFillOrderThrowIfInvalidAsync ` method so I feel this should be called before the `Promise.all`. This will also make this method cleaner, and remove the nested array returns of the `Promise.all`.